docs/architecture.md

```diff
@@ -0,0 +1,438 @@
+project, but frequently in Torch-MLIR this is the meaning that they have.
+Sometimes "frontend" can mean something even further up the stack, such as
+something in PyTorch itself. When there is ambiguity we will refer to this as
+"at the PyTorch level". Similarly, "backend" can sometimes refer to something
+sitting below Linalg-on-Tensors, TOSA, or MHLO.
+
+## The `torch` dialect
+
+See [include/torch-mlir/Dialect/Torch/IR](https://github.com/llvm/torch-mlir/tree/main/include/torch-mlir/Dialect/Torch/IR)
+
+The central MLIR abstraction in the Torch-MLIR project is the `torch` dialect.
+This dialect supports progressive lowering from the raw imported PyTorch
+programs that various PyTorch integration points provide, all the way down to
+the backend contract.
+
+The `torch` dialect must be versatile enough to support being imported by any
+program capture mechanism in PyTorch -- this could be TorchDynamo, `torch.fx`,
+LazyTensorCore, TorchScript, `torch.jit.trace`, etc. Thankfully, PyTorch is
+factored such that we can handle this with one core import path, which is
+through the PyTorch "[JIT
+IR](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md)",
+and lives in
+[torch-mlir/python/torch_mlir/dialects/torch/importer/jit_ir](https://github.com/llvm/torch-mlir/tree/e322f6a8784009b37aa354abfa9a40a80f30877d/python/torch_mlir/dialects/torch/importer/jit_ir).
+The JIT IR is a highly principled IR that faithfully models a Python subset (+
+tensors, the PyTorch op registry, and a few other things). All the other PyTorch
+program representations can eventually bottom-out on the JIT IR via some path
+provided by PyTorch. The `torch` dialect is almost entirely in 1:1
+correspondence with the JIT IR -- this allows the importer to be extremely small
+(the core is
+[under 500 lines of code](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/python/torch_mlir/dialects/torch/importer/jit_ir/csrc/node_importer.cpp#L1)).
+
+### Ops
+
+See [TorchOps.td](https://github.com/llvm/torch-mlir/blob/114f48e96c578ee76a6f83b3aa4aa229a8d5b76e/include/torch-mlir/Dialect/Torch/IR/TorchOps.td#L1)
+
+The ops in the `torch` dialect are almost entirely generated based on the
+PyTorch JIT IR operator registry via the script
+[torch_ods_gen.py](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L1) (invoked via [update_torch_ods.sh](https://github.com/llvm/torch-mlir/blob/main/build_tools/update_torch_ods.sh)).
+This script queries the registry and generates MLIR
+[ODS](https://mlir.llvm.org/docs/OpDefinitions/) in
+[GeneratedTorchOps.td](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td#L1). We have a guide for [adding a new op end-to-end](https://github.com/llvm/torch-mlir/wiki/Torch-ops-E2E-implementation).
```

Comment:
(can be addressed as a follow-on) Would it help to move this wiki page (and any other wiki page) as a Markdown doc tracked in code?

Reply:
Yes, I've been thinking the same thing.

Reply:
This dates to a much earlier phase of the project where we didn't have the same docs structure.